### PR TITLE
feat: Add option to match some of the labes instead of all #2122

### PR DESCRIPTION
**Pull request by sblack4**
Resolves https://github.com/philips-labs/terraform-aws-github-runner/issues/2122

Implements a feature that gives the option to have the webhook match [some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) of the labels rather than [every](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every) label. 

My current test case is two groups of GitHub runners, one arm and one amd64. I want to use a single tag to trigger the correct one. I can't share the whole thing but this terraform and simple action should be enough to convert what's going on: 

<details>
<summary> terraform </summary>

```

resource "random_id" "random" {
  byte_length = 20
}

data "aws_ssm_parameter" "app_id" {
  name = "/github_runner/app_id"
}

data "aws_ssm_parameter" "key_base64" {
  name = "/github_runner/key_base64"
}

module "runners" {
  source = "git::github.com/sblack4/terraform-aws-github-runner.git?ref=match-some-labes"

  ami_owners                                  = ["099720109477"] # Canonical's Amazon account ID
  aws_region                                  = var.region
  create_service_linked_role_spot             = true
  enable_organization_runners                 = true
  enable_ssm_on_runners                       = true
  fifo_build_queue                            = true
  instance_types                              = ["m5.large", "c5.large"]
  minimum_running_time_in_minutes             = 60
  runner_as_root                              = true
  runner_boot_time_in_minutes                 = 15
  runner_enable_workflow_job_labels_check     = true
  runner_enable_workflow_job_labels_check_all = false
  runner_extra_labels                         = "amd64,${var.env}"
  runner_iam_role_managed_policy_arns         = local.github_managed_policies
  subnet_ids                                  = local.private_subnets
  tags                                        = merge(local.tags, { "Name" = "github-runners" })
  userdata_template                           = "./userdata.sh"
  vpc_id                                      = local.vpc_id

  # lambdas for this version were manually downloaded
  runner_binaries_syncer_lambda_zip = "lambdas-download/runner-binaries-syncer.zip"
  runners_lambda_zip                = "lambdas-download/runners.zip"
  webhook_lambda_zip                = "lambdas-download/webhook.zip"

  ami_filter = {
    name = ["ubuntu/images/hvm-ssd/ubuntu-focal-20.04-amd64-server-*"]
  }

  block_device_mappings = [{
    delete_on_termination = true
    device_name           = "/dev/sda1"
    encrypted             = true
    iops                  = null
    volume_size           = 30
    volume_type           = "gp3"
  }]

  github_app = {
    key_base64     = data.aws_ssm_parameter.key_base64.value
    id             = data.aws_ssm_parameter.app_id.value
    webhook_secret = random_id.random.hex
  }

  idle_config = [{
    cron      = "* * 9-19 * * *"
    timeZone  = "America/Cancun"
    idleCount = 1
  }]

  # runner_binaries_s3_sse_configuration = {
  #   rule = {
  #     apply_server_side_encryption_by_default = {
  #       sse_algorithm = "AES256"
  #     }
  #   }
  # }

  runner_log_files = [
    {
      "log_group_name" : "syslog",
      "prefix_log_group" : true,
      "file_path" : "/var/log/syslog",
      "log_stream_name" : "{instance_id}"
    },
    {
      "log_group_name" : "user_data",
      "prefix_log_group" : true,
      "file_path" : "/var/log/user-data.log",
      "log_stream_name" : "{instance_id}/user_data"
    },
    {
      "log_group_name" : "runner",
      "prefix_log_group" : true,
      "file_path" : "/opt/actions-runner/_diag/Runner_**.log",
      "log_stream_name" : "{instance_id}/runner"
    }
  ]
}

resource "random_id" "random_arm" {
  byte_length = 20
}

data "aws_ssm_parameter" "app_id_arm" {
  name = "/github_runner/app_id_arm"
}

data "aws_ssm_parameter" "key_base64_arm" {
  name = "/github_runner/key_base64_arm"
}

module "arm_runners" {
  source = "git::github.com/sblack4/terraform-aws-github-runner.git?ref=match-some-labes"

  ami_owners                                  = ["099720109477"] # Canonical's Amazon account ID
  aws_region                                  = var.region
  create_service_linked_role_spot             = true
  enable_organization_runners                 = true
  enable_ssm_on_runners                       = true
  fifo_build_queue                            = true
  instance_types                              = ["t4g.large", "c6g.large"]
  minimum_running_time_in_minutes             = 60
  prefix                                      = "github-actions-arm"
  runner_architecture                         = "arm64"
  runner_as_root                              = true
  runner_boot_time_in_minutes                 = 15
  runner_enable_workflow_job_labels_check     = true
  runner_enable_workflow_job_labels_check_all = false
  runner_extra_labels                         = "arm64,${var.env}"
  runner_iam_role_managed_policy_arns         = local.github_managed_policies
  subnet_ids                                  = local.private_subnets
  tags                                        = merge(local.tags, { "Name" = "github-runners" })
  userdata_template                           = "./userdata_arm.sh"
  vpc_id                                      = local.vpc_id

  # lambdas for this version were manually downloaded
  runner_binaries_syncer_lambda_zip = "lambdas-download/runner-binaries-syncer.zip"
  runners_lambda_zip                = "lambdas-download/runners.zip"
  webhook_lambda_zip                = "lambdas-download/webhook.zip"

  ami_filter = {
    name = ["ubuntu/images/hvm-ssd/ubuntu-focal-20.04-arm64-server-*"]
  }

  block_device_mappings = [{
    delete_on_termination = true
    device_name           = "/dev/sda1"
    encrypted             = true
    iops                  = null
    volume_size           = 30
    volume_type           = "gp3"
  }]

  github_app = {
    key_base64     = data.aws_ssm_parameter.key_base64_arm.value
    id             = data.aws_ssm_parameter.app_id_arm.value
    webhook_secret = random_id.random_arm.hex
  }

  idle_config = [{
    cron      = "* * 9-19 * * *"
    timeZone  = "America/Cancun"
    idleCount = 1
  }]

  # runner_binaries_s3_sse_configuration = {
  #   rule = {
  #     apply_server_side_encryption_by_default = {
  #       sse_algorithm = "AES256"
  #     }
  #   }
  # }

  runner_log_files = [
    {
      "log_group_name" : "syslog",
      "prefix_log_group" : true,
      "file_path" : "/var/log/syslog",
      "log_stream_name" : "{instance_id}"
    },
    {
      "log_group_name" : "user_data",
      "prefix_log_group" : true,
      "file_path" : "/var/log/user-data.log",
      "log_stream_name" : "{instance_id}/user_data"
    },
    {
      "log_group_name" : "runner",
      "prefix_log_group" : true,
      "file_path" : "/opt/actions-runner/_diag/Runner_**.log",
      "log_stream_name" : "{instance_id}/runner"
    }
  ]
}
```
</details>

<details>
<summary> github action </summary>

```
---

name: test

on:
  push:

jobs:
  amd64:
    name: Build Docker
    runs-on: "x64"
    steps:
      - run: |
          pwd
          whoami
          env
          cat /etc/os-*
          uname -a

  arm64:
    name: Build Docker
    runs-on: "arm64"
    steps:
      - run: |
          pwd
          whoami
          env
          cat /etc/os-*
          uname -a
```

</details>

### CODE REVIEW

File: modules/webhook/lambdas/webhook/src/webhook/handler.ts [hunk 1]
```diff
@@ -171,10 +175,17 @@ function isRepoNotAllowed(repoFullName: string, repositoryWhiteList: string[]):
...
   const workflowJobLabels = job.workflow_job.labels;
-  const runnerMatch = runnerLabels.every((l) => workflowJobLabels.includes(l));
-  const jobMatch = workflowJobLabels.every((l) => runnerLabels.includes(l));
+  let runnerMatch;
+  let jobMatch
+  if(workflowLabelCheckAll) {
+    runnerMatch = runnerLabels.every((l) => workflowJobLabels.includes(l));
+    jobMatch = workflowJobLabels.every((l) => runnerLabels.includes(l));
+  } else {
+    runnerMatch = runnerLabels.some((l) => workflowJobLabels.includes(l));
```
Comment (npalm) [inline modules/webhook/lambdas/webhook/src/webhook/handler.ts:186]: Please can you add a unit test case for this else branch (some match)

**Review by npalm — COMMENTED** [review]
In the provide example you only pass labels like `arm64` or `x64` but yo not mention self-hosted for example. So just wondering what the use case is. When using GitHub.com this will also let job runs on public runners or any available matching runners.

File: modules/webhook/lambdas/webhook/src/webhook/handler.ts (hunk 1 above)
Comment (sblack4) [reply modules/webhook/lambdas/webhook/src/webhook/handler.ts:186]: The use-case is we have workflows with those labels (and just those labels) that we'd like to have run on one or the other group of autoscaling runners. I don't think there are any public runners or other runners that could match that label but I could be wrong?

Comment (npalm) [reply modules/webhook/lambdas/webhook/src/webhook/handler.ts:186]: Can you have a look on the linter that is failing, and add a test?

Comment (sblack4) [reply modules/webhook/lambdas/webhook/src/webhook/handler.ts:186]: Yup, give me a few days 🙏

Comment (sblack4) [reply modules/webhook/lambdas/webhook/src/webhook/handler.ts:187]: I added some tests and made the linter happy :)

**Review by npalm — APPROVED** [review]
@sblack4 thanks for your PR, tested with the default example. Seems all fine. I am planning in the future some serious refactory on the webhook. This too easu support multiple runnter types for a single app. Thinking about to move to a more filter appraoch. But for now this variable is fine!

**Pull request merged**